src/ossl/ecc.rs

```diff
@@ -709,3 +716,149 @@ impl Verify for EccOperation {
+            if self.mech == CKM_ECDH1_COFACTOR_DERIVE {
+                1
+            } else {
+                -1
+            } as i32,
+        )?;
+        let factory =
+            objfactories.get_obj_factory_from_key_template(template)?;
+
+        let keylen = match template.iter().find(|x| x.type_ == CKA_VALUE_LEN) {
+            Some(a) => a.to_ulong()? as usize,
+            None => {
+                let max = hash::hash_size(
+                    self.hash.ok_or(to_rv!(CKR_MECHANISM_INVALID))?,
+                );
+                if max == CK_UNAVAILABLE_INFORMATION as usize {
+                    return err_rv!(CKR_MECHANISM_INVALID);
+                }
+                match factory.as_secret_key_factory()?.recommend_key_size(max) {
+                    Ok(len) => len,
+                    Err(_) => return err_rv!(CKR_TEMPLATE_INCONSISTENT),
+                }
+            }
+        };
+        /* these do not apply to the raw ECDH */
+        if self.kdf != CKD_NULL {
+            params = params
+                .add_const_c_string(
+                    name_as_char(OSSL_EXCHANGE_PARAM_KDF_TYPE),
+                    ckd_to_type_name(self.kdf),
+                )?
+                .add_const_c_string(
+                    name_as_char(OSSL_EXCHANGE_PARAM_KDF_DIGEST),
+                    mech_type_to_digest_name(
+                        self.hash.ok_or(to_rv!(CKR_MECHANISM_INVALID))?,
+                    ),
+                )?;
+            if self.shared.len() > 0 {
+                params = params.add_octet_string(
+                    name_as_char(OSSL_EXCHANGE_PARAM_KDF_UKM),
+                    &self.shared.to_vec(),
+                )?;
+            }
+            params = params.add_uint(
+                name_as_char(OSSL_EXCHANGE_PARAM_KDF_OUTLEN),
+                keylen as u32,
+            )?;
+        }
+
+        params = params.finalize();
+
+        let mut pkey = object_to_ecc_private_key(key)?;
+        let mut ctx = new_pkey_ctx_from_obj(&mut pkey)?;
+        let res = unsafe {
+            EVP_PKEY_derive_init_ex(ctx.as_mut_ptr(), params.as_ptr())
+        };
+        if res != 1 {
+            return err_rv!(CKR_DEVICE_ERROR);
+        }
+
+        /* Import peer key */
+        let curve_name = get_curve_name_from_obj(key)?;
+        let public = self.public.to_vec();
+        let mut peer = make_ecc_public_key(curve_name, public)?;
+
+        let res = unsafe {
+            EVP_PKEY_derive_set_peer(ctx.as_mut_ptr(), peer.as_mut_ptr())
+        };
+        if res != 1 {
+            return err_rv!(CKR_DEVICE_ERROR);
+        }
+
+        let mut secret_len = 0usize;
+        let res = unsafe {
+            EVP_PKEY_derive(
+                ctx.as_mut_ptr(),
+                std::ptr::null_mut(),
+                &mut secret_len,
+            )
+        };
+        if res != 1 {
+            return err_rv!(CKR_DEVICE_ERROR);
+        }
+        let mut secret = vec![0u8; secret_len];
```

Comment:
change order between this one and the next?

Reply:
Will do.